python/ray/metrics_agent.py

```diff
@@ -0,0 +1,180 @@
+    @property
+    def units(self):
+        return self.measure.unit
+
+    @property
+    def tags(self):
+        return self.view.columns
+
+    def __dict__(self):
+        return {
+            "name": self.measure.name,
+            "description": self.measure.description,
+            "units": self.measure.unit,
+            "tags": self.view.columns,
+        }
+
+    def __str__(self):
+        return self.__repr__()
+
+    def __repr__(self):
+        return str(self.__dict__())
+
+
+class MetricsAgent:
+    def __init__(self, metrics_export_port):
+        if not metrics_export_port:
+            metrics_export_port = 8888
+
+        # OpenCensus classes.
+        self.stats = stats_module.stats
+        self.view_manager = self.stats.view_manager
+        self.stats_recorder = self.stats.stats_recorder
+        # Port where we will expose metrics.
+        self.metrics_export_port = metrics_export_port
+        # Measure to record metrics.
+        self._measure_map = self.stats_recorder.new_measurement_map()
+        # metric name(str) -> view (view.View)
+        # It is updated dynamically as new metrics
+        # are delivered from cpp processes.
+        self._registry = defaultdict(lambda: None)
+        # Lock required because gRPC server uses multi-threads.
+        self._lock = threading.Lock()
+        # Whether or not there are metrics that are missing description and
+        # units information. This is used to dynamically update registry.
+        self._missing_information = False
+
+        # Configure exporter. (We currently only support prometheus).
+        self.view_manager.register_exporter(
+            prometheus_exporter.new_stats_exporter(
+                prometheus_exporter.Options(
+                    namespace="ray", port=metrics_export_port)))
+
+    @property
+    def registry(self):
+        return self._registry
+
+    def record_metrics_points(self, metrics_points):
+        with self._lock:
+            for metric_point in metrics_points:
+                self._register_if_needed(metric_point)
+                self._record(metric_point)
+            return self._missing_information
+
+    def _record(self, metric_point):
+        """Record a single metric point to export.
+
+        NOTE: When this method is called, the caller should acquire a lock.
+
+        Args:
+            metric_point metric point defined in common.proto
+        """
+        metric_name = metric_point.metric_name
+        tags = metric_point.tags
+
+        metric = self._registry.get(metric_name)
+        # Metrics should be always registered dynamically.
+        assert metric
+
+        tag_map = tag_map_module.TagMap()
+        for key, value in tags.items():
+            tag_key = tag_key_module.TagKey(key)
+            tag_value = tag_value_module.TagValue(value)
+            tag_map.insert(tag_key, tag_value)
+
+        metric_value = metric_point.value
+        # The class should be thread-safe because GRPC server
```

Comment:
I'm not sure what this comment means - what class should be thread safe? Aren't we holding a lock here anyways?

Reply:
Just removed. Agree the comment was weird (It was written before I added a lock on the upper level. )